Fully Implement Resource Templates

Tonksthebear opened the pull request:
Hello! I stumbled upon this project and wanted to start contributing if you're interested.

This PR:

- Implements Resource Templates Fully
- Enhances the `read_resource` method to search for corresponding resource templates in the event no hard coded resource is found
- Adds corresponding tests

It also
- Adds the manually specified `version` to the list of valid supported protocol versions (some MCP clients request a version unrelated to protocol version, like Goose)
- Makes cursor `nil` instead of an empty string when there are no paginated pages. Certain clients (like goose) think all non-nils are valid links and end up in an endless cycle of requesting "". `nil` solves this

Those last two additions can be split into a separate PR if you'd like (and if you're interested in them at all). Also, please feel free to nitpick this code, I'm happy to change it as you see fit.

funwarioisii commented on lib/mcp/app/tool.rb:
```diff
@@ -77,7 +77,7 @@ def list_tools(cursor: nil, page_size: 10)
         start_index = cursor ? cursor.to_i : 0
         paginated = tool_values[start_index, page_size]
-        next_cursor = (start_index + page_size < tool_values.length) ? (start_index + page_size).to_s : ""
+        next_cursor = (start_index + page_size < tool_values.length) ? (start_index + page_size).to_s : nil
```
I checked again, and it looks like `nil` is correct.

https://github.com/modelcontextprotocol/specification/blob/6828f3ef6300b25dd2aaff2a2e5e81188bdbd22e/schema/2024-11-05/schema.ts#L306-L312

funwarioisii commented on lib/mcp/app/resource_template.rb:
```diff
@@ -0,0 +1,142 @@
+        else
+          paginated = values[start_index, page_size]
+          has_next = start_index + page_size < values.length
+          next_cursor = has_next ? (start_index + page_size).to_s : ""
```
I think it would be good to align this with `list_tools` and set it to `nil` as well.

https://github.com/modelcontextprotocol/specification/blob/6828f3ef6300b25dd2aaff2a2e5e81188bdbd22e/schema/2024-11-05/schema.ts#L332-L334

funwarioisii reviewed (commented):
Amazing! I believe this is an essential Pull Request for the project.
I really appreciate that you’ve also addressed some fixes that hadn’t been covered yet.

Could you please check the following points?

- I think it would be good to add a test to verify that multiple arguments can be expanded in the template using /. What do you think?
- This is my fault for not setting up CI, but could you format the code using `bundle exec standardrb --fix`?

I think the test could look something like this:

```ruby
      def test_extract_variables
        @app.register_resource_template("/test/{param1}/{param2}") do
          name "test_template"
          call { |args| "#{args[:param1]}, #{args[:param2]}" }
        end

        template, values = @app.find_matching_template("/test/value1/value2")
        refute_nil template
        assert_equal({ param1: "value1", param2: "value2" }, values)
      end
```

What do you think?

funwarioisii commented:
I forgot to mention you, so I'm tagging you again. @Tonksthebear

Tonksthebear commented:
@funwarioisii I updated next_cursor to be nil in a couple of extra spots (overlooked that in the original implementation)

I also added your suggested test. I thought I already had a multi-param test, but alas I did not 😆 I also ran standardrb

funwarioisii commented on lib/mcp/app/resource_template.rb:
```diff
@@ -0,0 +1,143 @@
+
+        # Creates a pattern for matching URIs against this template
+        def to_pattern
+          pattern_string = Regexp.escape(@uri_template).gsub(/\\\{[^}]+\\\}/) do |match|
+            match.gsub(/\\\{|\\\}/, "")
+            "([^/]+)"
+          end
```
What is these lines expected to do?
`match.gsub(/\\\{|\\\}/, "")`

Tonksthebear commented:
Ugh sorry about that, I didn't commit its removal. It was meant to be removed. Now it is

funwarioisii reviewed (approved):
Thank you so much for the fantastic Pull Request!
I'll publish this gem, including this change, soon.

funwarioisii commented:
Published! https://rubygems.org/gems/mcp-rb
FYI: @Tonksthebear

Tonksthebear commented:
Awesome, thank you so much @funwarioisii !
